detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js

```diff
@@ -0,0 +1,66 @@
+const _ = require('lodash');
+const AndroidDriver = require('../AndroidDriver');
+const GenyCloudExec = require('./exec/GenyCloudExec');
+const RecipesService = require('./services/GenyRecipesService');
+const InstanceLookupService = require('./services/GenyInstanceLookupService');
+const InstanceLifecycleService = require('./services/GenyInstanceLifecycleService');
+const InstanceNaming = require('./services/GenyInstanceNaming');
+const AllocationHelper = require('./helpers/GenyAllocationHelper');
+const DeviceQueryHelper = require('./helpers/GenyDeviceQueryHelper');
+const DetoxRuntimeError = require('../../../../errors/DetoxRuntimeError');
+const logger = require('../../../../utils/logger').child({ __filename });
+
+class GenyCloudDriver extends AndroidDriver {
+  constructor(config) {
+    super(config);
+    this._name = 'Unspecified Genymotion Cloud Emulator';
+
+    const exec = new GenyCloudExec();
+    const instanceNaming = new InstanceNaming(); // TODO should consider a permissive impl for debug/dev mode. Maybe even a custom arg in package.json (Detox > ... > genycloud > sharedAccount: false)
+    this.recipeService = new RecipesService(exec, logger);
+    this.instanceLookupService = new InstanceLookupService(exec, instanceNaming, this.deviceRegistry);
+    this.instanceLifecycleService = new InstanceLifecycleService(exec, instanceNaming);
+    this.deviceQueryHelper = new DeviceQueryHelper(this.recipeService);
+    this.allocationHelper = new AllocationHelper(this.instanceLookupService, this.instanceLifecycleService);
+  }
+
+  get name() {
+    return this._name;
+  }
+
+  async acquireFreeDevice(deviceQuery) {
+    const recipe = await this.deviceQueryHelper.getRecipeFromQuery(deviceQuery);
+    this.assertRecipe(deviceQuery, recipe);
```

Comment:
**No, you are not getting a recipe here.**

Here you are just resolving a "maybe ambiguous device query", where it can be deterministic (recipe UUID) or fuzzy (recipe name).

You get a guaranteed elaborate recipe object only when you work with `gmsaas instances`. Your stub type for recipe confuses me a lot, to be honest, looking more like a temporary workaround inside of redesigning a correct flow of API calls and handling errors in a most user-friendly fashion.

Reply:
I don't understand what you mean with that... `deviceQueryHelper` _does_ return a receipe. `recipesService.getRecipeByName` Does fetch for a recipe. What's misleading here?

Reply:
Relates to #2575